### PR TITLE
Store unknown fields in ProtoReader instead of Builder

### PR DESCRIPTION
**Pull request by Egorand**

### CODE REVIEW

File: wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoReader.kt [hunk 1]
```diff
@@ -90,13 +93,31 @@ class ProtoReader(private val source: BufferedSource) {
...
+  }
+
+  /**
+   * End a length-delimited nested message. Calls to this method must be symmetric with calls to
+   * [beginMessage].
+   *
+   * @param token value returned from the corresponding call to [beginMessage].
+   */
+  @Throws(IOException::class)
+  @Deprecated(level = DeprecationLevel.HIDDEN, message = "prefer endMessageAndGetUnknownFields()")
```
Comment (swankjesse) [inline wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoReader.kt]: Wanna do WARNING deprecation? Code that has a compiler/runtime mismatch should still build?

File: wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoReader.kt [hunk 2]
```diff
@@ -391,29 +412,22 @@ class ProtoReader(private val source: BufferedSource) {
...
-      @Suppress("UNCHECKED_CAST") // We encode and decode the same types.
-      (protoAdapter as ProtoAdapter<Any>).encodeWithTag(unknownFieldsWriter!!, tag, value)
+      tagHandler(tag)
     }
-    endMessage(token)
+    return endMessageAndGetUnknownFields(token)
+  }
 
-    return unknownFieldsBuffer?.readByteString() ?: ByteString.EMPTY
+  fun readUnknownField(tag: Int) {
```
Comment (swankjesse) [inline wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoReader.kt]: Document that the unknown field will be included in endMessageAndGetUnknownFields’s return value?

**Comment by swankjesse** [comment]
Worth adding a test where the enclosing and enclosed messages both have unknown fields? As is I think we could incorrectly share buffers and our tests would still pass.

File: wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoReader.kt (hunk 1 above)
Comment (Egorand) [reply wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoReader.kt]: Good catch, code generated by previous versions of compiler will fail to compile after this change.

**Comment by Egorand** [comment]
@swankjesse PTAL https://github.com/square/wire/pull/1017/commits/dc36207d177a0e61c9847e6eeafe8282c3f5d108

**Pull request merged**